sympy/assumptions/cnf.py

```diff
@@ -0,0 +1,373 @@
+from itertools import combinations, product
+
+from sympy import S
+from sympy.logic.boolalg import BooleanFunction, Or, And, Not
+
+
+# Early encoding for literals
+literals_store = dict()
+
+
+class Literal(object):
+    """
+    The smallest element of a CNF object
+    """
+    literals_count = 0
+
+    def __new__(cls, lit, is_not=False):
+        if type(lit).__name__ == 'Not':
```

Comment:
Use isinstance here and elsewhere instead of comparing type strings.

Reply:
Out of everything I tried this is the fastest. `isinstance()` is better when something is going to inherit `Not` and other classes used. I don't think it's going to be a case here.

Reply:
Does it actually affect the end runtime, though? I tested and they both take on the order of 100 ns.

Reply:
The advantage comes when the object is not a `Not` object. Though the difference appears to be marginal and I can't say I understand how `isinstance` works but I think it goes into the inheritance of the object also.

Reply:
Normally `isinstance(obj, cls)` is equivalent to `cls in type(obj).mro()` but it is also possible to override this with metaclasses that define `__instancecheck__`.

There are a couple of problems with what you have here. One is that an unrelated type that happens to have the same name will pass this check because you're just checking the name string. You could fix that by comparing the type objects with `type(lit) is Not` but that also suffers from the other problem which is that the test wouldn't recognise instances of subclasses of Not.

Unless you have a strong (and unusual) reason for doing anything different you should just use isinstance and not think about it. The big gains here are coming from higher-level changes so you should focus on implementing those as cleanly as possible rather than on micro-optimisations like trying to speed up isinstance.